scripts/pages/zoning/zoning.ts

```diff
@@ -813,218 +1005,185 @@ class ZoneMenuHandler {
+	setInfoPanelShown(show: boolean) {
+		this.panels.propertiesPanel.SetHasClass('hide', show);
+		this.panels.infoPanel.SetHasClass('hide', !show);
+
+		if (!show) {
+			this.populateRegionProperties();
+		}
+	}
+
+	createMain() {
+		if (!this.mapZoneData || Boolean(this.mapZoneData.tracks.main)) return;
+
+		this.mapZoneData.tracks.main = {
+			zones: {
+				segments: [this.createSegment()],
+				end: {}
+			},
+			stagesEndAtStageStarts: true
+		};
+
+		this.updateSelection({
+			track: this.mapZoneData.tracks.main,
+			segment: this.mapZoneData.tracks.main.zones.segments[0],
+			zone: this.mapZoneData.tracks.main.zones.segments[0].checkpoints[0]
+		});
 
-		if (!hideProperties) this.populateRegionProperties();
+		this.panels.zoningMenu.createRegion(true);
+		this.setInfoPanelShown(true);
 	}
 
-	addBonus() {
+	addBonus(defragBonus = false, index = null) {
 		if (!this.mapZoneData) return;
 
-		const lastTrack = this.panels.trackList.GetChild(this.mapZoneData.tracks.bonuses?.length ?? 0)!;
-		const addBonusButton = lastTrack.FindChildTraverse<Panel>('AddBonusButton')!;
-		addBonusButton.SetHasClass('hide', true);
+		const bonus = defragBonus ? { defragModifiers: 1 } : this.createBonusTrack();
 
-		const bonus = this.createBonusTrack();
 		if (!this.mapZoneData.tracks.bonuses) {
-			this.mapZoneData.tracks.bonuses = [bonus];
+			this.mapZoneData.tracks.bonuses = [];
+		}
+
+		if (index != null && index < this.mapZoneData.tracks.bonuses.length) {
+			this.mapZoneData.tracks.bonuses.splice(Math.max(index, 0), 0, bonus);
 		} else {
 			this.mapZoneData.tracks.bonuses.push(bonus);
 		}
 
-		this.createTrackEntry(
-			this.panels.trackList,
-			bonus,
-			`${$.Localize('#Zoning_Bonus')} ${this.mapZoneData.tracks.bonuses.length}`
-		);
-
-		const newBonusPanel = this.panels.trackList.GetChild(this.mapZoneData.tracks.bonuses.length)!;
-		const segmentContainer = newBonusPanel.FindChildTraverse('SegmentContainer')!;
-		const checkpointContainer = segmentContainer.FindChildTraverse<Panel>('CheckpointContainer')!;
-		this.addCheckpoint(bonus, bonus.zones!.segments[0], checkpointContainer);
-
-		const selectButton = checkpointContainer.FindChildTraverse<ToggleButton>('SelectButton')!;
-		const deleteButton = checkpointContainer.FindChildTraverse<Button>('DeleteButton')!;
-		const newBonusButton = newBonusPanel.FindChildTraverse<Panel>('AddBonusButton')!;
-		newBonusButton.SetHasClass('hide', false);
-		selectButton.SetSelected(true);
+		if (defragBonus) {
+			this.updateSelection({ track: bonus });
+			this.showDefragFlagMenu();
+		} else {
+			this.updateSelection({
+				track: bonus,
+				segment: bonus.zones.segments[0],
+				zone: bonus.zones.segments[0].checkpoints[0]
+			});
 
-		this.updateSelection(bonus, bonus.zones!.segments[0], bonus.zones!.segments[0].checkpoints[0], deleteButton);
+			this.panels.zoningMenu.createRegion(true);
+			this.setInfoPanelShown(true);
+		}
 	}
 
 	addSegment() {
```

Comment:
gave you snippets for the other portions of this, but left this one out. This is what I was playing with
```js
addSegment(index = null) {
	if (!this.mapZoneData) return;
	if (!this.selectedZone.track || !this.isMainTrack(this.selectedZone.track)) return;

	const newSegment = this.createSegment();

	if (!this.selectedZone.track.zones.segments) {
		this.selectedZone.track.zones.segments = [];
	}

	this.insertItem(this.selectedZone.track.zones.segments, index, newSegment);

	this.updateSelection({ track: this.selectedZone.track, segment: newSegment, zone: newSegment.checkpoints[0] });

	this.panels.zoningMenu.createRegion(true);
	this.setInfoPanelShown(true);
}
```

Reply:
I've added a context menu for inserting segments. I also added checks so each context menu is disabled if you are not allowed to create more of that thing for whatever reason.

Reply:
nice, that sounds good to me